### PR TITLE
Feature/local cache

### PR DESCRIPTION
**Pull request by Boukhari-Mohamed**

### CODE REVIEW

**Review by raedbm — APPROVED** [review]
The cache/offline feature looks good. I didn't get into details when looking into the pull request but everything looks fine.

File: app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/cache/CacheDatabase.java [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+            return (T)meal;
+        }
+    }
+
+    private class CollectionSnapshotImpl implements CollectionSnapshot{
+        private List<DocumentSnapshot> documents;
+        @SuppressLint("NewApi")
+        CollectionSnapshotImpl(Collection<DocumentSnapshot> documents){
+            this.documents = documents.stream().collect(toList());
+        }
```
Comment (khalilacheche) [inline app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/cache/CacheDatabase.java:107]: it's better to use another way without having a warning on the api

File: app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/cache/CacheDatabase.java [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+        return instance;
+    }
+
+
+    public void reset() {
+        initializeFields();
+    }
+
+    public void persist() throws Exception{
```
Comment (khalilacheche) [inline app/src/main/java/com/epfl/neighborfood/neighborfoodandroid/database/cache/CacheDatabase.java:155]: how is persistance done with the firebase database?

**Review by khalilacheche — APPROVED** [review]
Good implementation overall, however it feels like some features are missing, like doing the persistance (I.e. the method call), and also a nice feature would have been to add a listener on the documents that are cached, so that we make sure we don't have a stale version of the documents restored to the users.

**Pull request merged**